src/services/identity/UsersService.ts

```diff
@@ -188,26 +210,106 @@ class UsersService {
+
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({ where: { email } })
+    if (!otpEntry) {
+      // create new entry
+      await this.createOtpEntry(email, OtpType.Email, hashedOtp)
+    } else {
+      await otpEntry?.update({
+        hashedOtp,
+        attempts: 0,
+        expiresAt: this.getOtpExpiry(),
+      })
+    }
 
     const subject = "One-Time Password (OTP) for IsomerCMS"
-    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${expiry} minutes. Please use this to verify your email address.</p>
+    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your email address.</p>
     <p>If your OTP does not work, please request for a new OTP.</p>
     <p>IsomerCMS Support Team</p>`
     await this.mailer.sendMail(email, subject, html)
   }
 
   async sendSmsOtp(mobileNumber: string) {
-    const otp = this.otp.generate(mobileNumber)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
 
-    const message = `Your OTP is ${otp}. It will expire in ${expiry} minutes. Please use this to verify your mobile number`
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({
+      where: { mobileNumber },
+    })
+    if (!otpEntry) {
+      await this.createOtpEntry(mobileNumber, OtpType.Mobile, hashedOtp)
+    } else {
+      await otpEntry?.update({ hashedOtp, attempts: 0 })
+    }
+
+    const message = `Your OTP is ${otp}. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your mobile number`
     await this.smsClient.sendSms(mobileNumber, message)
   }
 
-  verifyOtp(value: string, otp: string) {
-    return this.otp.verify(value, otp)
+  async verifyOtp(key: string, keyType: OtpType, otp: string) {
+    let otpEntry
+    if (keyType === OtpType.Email) {
+      otpEntry = await this.otpRepository.findOne({ where: { email: key } })
+    } else {
+      otpEntry = await this.otpRepository.findOne({
+        where: { mobileNumber: key },
+      })
+    }
+
+    if (!otpEntry?.hashedOtp) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Hashed OTP not found")
+    }
+
+    if (
+      otpEntry.attempts !== null &&
+      otpEntry.attempts >= MAX_NUM_OTP_ATTEMPTS
+    ) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Max number of attempts reached")
+    }
+
+    // increment attempts
+    await otpEntry.update({ attempts: otpEntry.attempts + 1 })
+
+    const isValidOtp = await this.otpService.verifyOtp(otp, otpEntry.hashedOtp)
+    if (!isValidOtp) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("OTP is not valid")
+    }
+
+    return true
```

Comment:
we still seem to maintain the entry? so if the user logs in successfully -> token expires -> tries to log in again, they won't be able to cos the entry is still present w/ unique constraint

Reply:
everytime we generate a new otp, we update otp entry in the DB to have new Otp + reset attempts to 0

Reply:
```
// Reset attempts to login
    const otpEntry = await this.otpRepository.findOne({ where: { email } })
    if (!otpEntry) {
      // create new entry
      await this.createOtpEntry(email, OtpType.Email, hashedOtp)
    } else {
      await otpEntry?.update({
        hashedOtp,
        attempts: 0,
        expiresAt: this.getOtpExpiry(),
      })
    }
```

Reply:
yeap you're right! should just `destroy` the otp entry when login is successful tbh cos otherwise it'll still be valid

Reply:
@seaerchin okay added destroy after checking isValidOtp